### ISSUE TITLE
Got negative train loss when do pretrain process

### ISSUE THREAD

**Issue by Jingyao711** [MAGICS-LAB/DNABERT_2#97]
I am using the run_mlm.py with bert_layers.py. When I run run_mlm.py, I encountered an error: Ask me to install Transformer with version more than 4.42.0.dev0. Since the version in requirement.txt is transformers==4.29.2. I muted the check code:
![1719341318720](https://github.com/MAGICS-LAB/DNABERT_2/assets/85922017/5f7c62a3-d111-4283-b386-f500c036a42d)

In the script it want to install is_torch_xla_available but with the environment I have, I cannot install it. So I just commented the line of code:
![1719341474159](https://github.com/MAGICS-LAB/DNABERT_2/assets/85922017/2c5bfa5e-6207-493f-bae7-9bbc9037f489)

After run the script, I got a very weird train loss line:

![1719341880386](https://github.com/MAGICS-LAB/DNABERT_2/assets/85922017/c0d03db3-c9d9-444d-857b-f8e23d6440d9)

Can you help me to understand what might cause this error?

**Comment by Zhihan1996** [comment]
Are you using the same training data as DNABERT-2 and the same training hyperparameters? Can you share more details on your training?

**Comment by Jingyao711** [comment]
I am using the same 

> Are you using the same training data as DNABERT-2 and the same training hyperparameters? Can you share more details on your training?

I am using the same data. This is the data I downloaded from the link in pretrain section.
![1721059613064](https://github.com/user-attachments/assets/8e98b008-f801-4bae-a2bc-36af97fdf9dd)
Then I ran python run_mlm.py from https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling.
This is the train argument I set for the trainning:
06/10/2024 23:26:50 - WARNING - __main__ - Process rank: 0, device: cuda:0, n_gpu: 2, distributed training: False, 16-bits training: False
06/10/2024 23:26:50 - INFO - __main__ - Training/evaluation parameters TrainingArguments(
_n_gpu=2,
adafactor=False,
adam_beta1=0.9,
adam_beta2=0.999,
adam_epsilon=1e-08,
auto_find_batch_size=False,
bf16=False,
bf16_full_eval=False,
data_seed=None,
dataloader_drop_last=False,
dataloader_num_workers=0,
dataloader_pin_memory=True,
ddp_backend=None,
ddp_bucket_cap_mb=None,
ddp_find_unused_parameters=None,
ddp_timeout=1800,
debug=[],
deepspeed=None,
disable_tqdm=False,
do_eval=True,
do_predict=False,
do_train=True,
eval_accumulation_steps=None,
eval_delay=0,
eval_steps=None,
evaluation_strategy=no,
fp16=False,
fp16_backend=auto,
fp16_full_eval=False,
fp16_opt_level=O1,
fsdp=[],
fsdp_config={'fsdp_min_num_params': 0, 'xla': False, 'xla_fsdp_grad_ckpt': False},
fsdp_min_num_params=0,
fsdp_transformer_layer_cls_to_wrap=None,
full_determinism=False,
gradient_accumulation_steps=1,
gradient_checkpointing=False,
greater_is_better=None,
group_by_length=False,
half_precision_backend=auto,
hub_model_id=None,
hub_private_repo=False,
hub_strategy=every_save,
hub_token=<HUB_TOKEN>,
ignore_data_skip=False,
include_inputs_for_metrics=False,
jit_mode_eval=False,
label_names=None,
label_smoothing_factor=0.0,
learning_rate=5e-05,
length_column_name=length,
load_best_model_at_end=False,
local_rank=0,
log_level=passive,
log_level_replica=warning,
log_on_each_node=True,
logging_dir=/tmp/dnabert2-mlm/runs/Jun10_23-26-50_doublerainbow,
logging_first_step=False,
logging_nan_inf_filter=True,
logging_steps=500,
logging_strategy=steps,
lr_scheduler_type=linear,
max_grad_norm=1.0,
max_steps=-1,
metric_for_best_model=None,
mp_parameters=,
no_cuda=False,
num_train_epochs=3.0,
optim=adamw_hf,
optim_args=None,
output_dir=/tmp/dnabert2-mlm,
overwrite_output_dir=False,
past_index=-1,
per_device_eval_batch_size=8,
per_device_train_batch_size=8,
prediction_loss_only=False,
push_to_hub=False,
push_to_hub_model_id=None,
push_to_hub_organization=None,
push_to_hub_token=<PUSH_TO_HUB_TOKEN>,
ray_scope=last,
remove_unused_columns=True,
report_to=['wandb'],
resume_from_checkpoint=None,
run_name=/tmp/dnabert2-mlm,
save_on_each_node=False,
save_safetensors=False,
save_steps=500,
save_strategy=steps,
save_total_limit=None,
seed=42,
sharded_ddp=[],
skip_memory_metrics=True,
tf32=None,
torch_compile=False,
torch_compile_backend=None,
torch_compile_mode=None,
torchdynamo=None,
tpu_metrics_debug=False,
tpu_num_cores=None,
use_ipex=False,
use_legacy_prediction_loop=False,
use_mps_device=False,
warmup_ratio=0.0,
warmup_steps=0,
weight_decay=0.0,
xpu_backend=None,
)
Using custom data configuration default-a746955ea34eb632
06/10/2024 23:26:50 - INFO - datasets.builder - Using custom data configuration default-a746955ea34eb632
Loading Dataset Infos from /home/tangji19/miniconda3/envs/dnabert2/lib/python3.8/site-packages/datasets/packaged_modules/text
06/10/2024 23:26:50 - INFO - datasets.info - Loading Dataset Infos from /home/tangji19/miniconda3/envs/dnabert2/lib/python3.8/site-packages/datasets/packaged_modules/text
Overwrite dataset info from restored data version if exists.
06/10/2024 23:26:50 - INFO - datasets.builder - Overwrite dataset info from restored data version if exists.
Loading Dataset info from /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101
06/10/2024 23:26:50 - INFO - datasets.info - Loading Dataset info from /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101
Found cached dataset text (/home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101)
06/10/2024 23:26:51 - INFO - datasets.builder - Found cached dataset text (/home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101)
Loading Dataset info from /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101
06/10/2024 23:26:51 - INFO - datasets.info - Loading Dataset info from /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101
[WARNING|configuration_utils.py:605] 2024-06-10 23:30:05,814 >> The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
/home/tangji19/miniconda3/envs/dnabert2/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
[INFO|configuration_utils.py:669] 2024-06-10 23:30:05,892 >> loading configuration file config.json from cache at /home/tangji19/.cache/huggingface/hub/models--zhihan1996--DNABERT-2-117M/snapshots/d064dece8a8b41d9fb8729fbe3435278786931f1/config.json
[INFO|configuration_utils.py:725] 2024-06-10 23:30:05,892 >> Model config BertConfig {
  "_name_or_path": "zhihan1996/DNABERT-2-117M",
  "alibi_starting_size": 512,
  "attention_probs_dropout_prob": 0.0,
  "auto_map": {
    "AutoConfig": "zhihan1996/DNABERT-2-117M--configuration_bert.BertConfig",
    "AutoModel": "zhihan1996/DNABERT-2-117M--bert_layers.BertModel",
    "AutoModelForMaskedLM": "zhihan1996/DNABERT-2-117M--bert_layers.BertForMaskedLM",
    "AutoModelForSequenceClassification": "zhihan1996/DNABERT-2-117M--bert_layers.BertForSequenceClassification"
  },
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "torch_dtype": "float32",
  "transformers_version": "4.29.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 4096
}

[INFO|tokenization_utils_base.py:1810] 2024-06-10 23:30:05,933 >> loading file tokenizer.json from cache at /home/tangji19/.cache/huggingface/hub/models--zhihan1996--DNABERT-2-117M/snapshots/d064dece8a8b41d9fb8729fbe3435278786931f1/tokenizer.json
[INFO|tokenization_utils_base.py:1810] 2024-06-10 23:30:05,933 >> loading file added_tokens.json from cache at None
[INFO|tokenization_utils_base.py:1810] 2024-06-10 23:30:05,933 >> loading file special_tokens_map.json from cache at None
[INFO|tokenization_utils_base.py:1810] 2024-06-10 23:30:05,933 >> loading file tokenizer_config.json from cache at /home/tangji19/.cache/huggingface/hub/models--zhihan1996--DNABERT-2-117M/snapshots/d064dece8a8b41d9fb8729fbe3435278786931f1/tokenizer_config.json
[INFO|modeling_utils.py:2516] 2024-06-10 23:30:05,997 >> loading weights file pytorch_model.bin from cache at /home/tangji19/.cache/huggingface/hub/models--zhihan1996--DNABERT-2-117M/snapshots/d064dece8a8b41d9fb8729fbe3435278786931f1/pytorch_model.bin
/home/tangji19/.cache/huggingface/modules/transformers_modules/zhihan1996/DNABERT-2-117M/d064dece8a8b41d9fb8729fbe3435278786931f1/bert_layers.py:126: UserWarning: Unable to import Triton; defaulting MosaicBERT attention implementation to pytorch (this will reduce throughput when using this model).
  warnings.warn(
[WARNING|modeling_utils.py:3175] 2024-06-10 23:30:06,351 >> Some weights of the model checkpoint at zhihan1996/DNABERT-2-117M were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.predictions.decoder.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
[WARNING|modeling_utils.py:3187] 2024-06-10 23:30:06,351 >> Some weights of BertModel were not initialized from the model checkpoint at zhihan1996/DNABERT-2-117M and are newly initialized: ['bert.pooler.dense.weight', 'bert.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
06/10/2024 23:30:06 - WARNING - __main__ - The chosen tokenizer supports a `model_max_length` that is longer than the default `block_size` value of 1024. If you would like to use a longer `block_size` up to `tokenizer.model_max_length` you can override this default with `--block_size xxx`.

Running tokenizer on every text in dataset:   0%|          | 0/32387832 [00:00<?, ? examples/s]Caching processed dataset at /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101/cache-615faf5bc4fb980d.arrow
06/10/2024 23:30:06 - INFO - datasets.arrow_dataset - Caching processed dataset at /home/tangji19/.cache/huggingface/datasets/text/default-a746955ea34eb632/0.0.0/96636a050ef51804b84abbfd4f4ad440e01153c24b86293eb5c3b300a41f9101/cache-615faf5bc4fb980d.arrow

Running tokenizer on every text in dataset:   0%|          | 1000/32387832 [00:00<1:34:11, 5731.15 examples/s]
Running tokenizer on every text in dataset:   0%|          | 3000/32387832 [00:00<47:48, 11288.76 examples/s] 
Running tokenizer on every text in dataset:   0%|          | 5000/32387832 [00:00<39:55, 13519.62 examples/s]
Running tokenizer on every text in dataset:   0%|          | 7000/32387832 [00:00<36:43, 14694.61 examples/s]
Running tokenizer on every text in dataset:   0%|          | 9000/32387832 [00:00<34:55, 15455.25 examples/s]

**Comment by Zhihan1996** [comment]
Do you want to pre-train a model from scratch or do further pre-training on DNABERT-2 checkpoint? Looks like you are starting from the DNABERT-2 checkpoint.
